### PR TITLE
fix java source lang record handling

### PR DESCRIPTION
**Pull request by cdietrich**

### CODE REVIEW

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java [hunk 1]
```diff
@@ -126,6 +126,9 @@ public JvmDeclaredType createType(TypeDeclaration type, String packageName) {
 		case TypeDeclaration.ANNOTATION_TYPE_DECL:
 			jvmType = TypesFactory.eINSTANCE.createJvmAnnotationType();
 			break;
+		case 5 /*TypeDeclaration.RECORD_DECL*/:
```
Comment (cdietrich) [inline org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: see also discussion in

https://github.com/eclipse/xtext/issues/2823
and 
https://github.com/eclipse/xtext/pull/3037

Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: should we use the preview constant? it seems to be in 33 too

File: org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java [hunk 2]
```diff
@@ -217,9 +213,19 @@ public void testJava21Record() {
 
 	@Inject
 	private IJvmTypeProvider.Factory typeProviderFactory;
-
+
```
Comment (LorenzoBettini) [inline org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java:216]: Nip: line with only tab

File: org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java [hunk 3]
```diff
@@ -217,9 +213,19 @@ public void testJava21Record() {
...
+	
 	protected XtextResourceSet resourceSet(Map<String, String> files) {
+		return this.resourceSet(files, null);
+	}
+
+	protected XtextResourceSet resourceSet(Map<String, String> files, JavaVersion javaVersion) {
 		XtextResourceSet result = resourceSetProvider.get();
+		if (javaVersion != null) {
+			JavaConfig javaConfig = new JavaConfig();
+			javaConfig.setJavaSourceLevel(JavaVersion.JAVA21);
```
Comment (LorenzoBettini) [inline org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java:225]: I guess here you meant to pass `javaVersion`, the parameter, not the hardcoded version

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java (hunk 1 above)
Comment (LorenzoBettini) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: I would use `TypeDeclaration.RECORD_DECL`: we already use other internal constants; it should not be a preview feature anymore in Java 21; maybe they forgot to remove the comment? Or maybe they refer to previous Java versions.

I don't understand the other comment. You're not talking about specifically about records, but in general about the `default` case? Maybe it's better to default to generic type as we do in JdtBasedTypeFactory, though I don't understand the part about super type.

**Review by LorenzoBettini — COMMENTED** [review]
In general, I'd like to have a test that would fail without this change. Instead, `org.eclipse.xtext.java.tests.JavaSourceLanguageTest.testJava21Record()` still succeeds without this change.

For example, in the current shape, if I comment

```java
		javaConfig.setJavaSourceLevel(JavaVersion.JAVA21);
		javaConfig.setJavaTargetLevel(JavaVersion.JAVA21);
```

the test fails; so I was misled that everything else would work by just setting the Java version.

Or did I miss something?

See also the other small changes.

**Comment by cdietrich** [comment]
grafik

**Comment by cdietrich** [comment]
for me it fails without change

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java (hunk 1 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: the jdtbasedfactory does not throw. it falls back to jvm generic type

File: org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.java.tests/src/org/eclipse/xtext/java/tests/JavaSourceLanguageTest.java:225]: yes

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java (hunk 1 above)
Comment (cdietrich) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java]: i am asking if we should do it here to. any idea @szarnekow

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java [hunk 4]
```diff
@@ -126,6 +126,9 @@ public JvmDeclaredType createType(TypeDeclaration type, String packageName) {
 		case TypeDeclaration.ANNOTATION_TYPE_DECL:
 			jvmType = TypesFactory.eINSTANCE.createJvmAnnotationType();
 			break;
+		case TypeDeclaration.RECORD_DECL:
```
Comment (cdietrich) [inline org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: its a final. so it will get inlined anyway. afaik

**Comment by LorenzoBettini** [comment]
> for me it fails without change

@cdietrich 

Yes, I verified that in your branch.

I was referring to the previous shape of the test: without the JAVA_21 specification it was failing, with the specification it was succeeding, but it wasn't testing much as you said.

Maybe it's worthwhile to leave a comment in the `resourceSet` method in the test, or in the test class in general stressing that to correctly test something in this test you should... etc. etc.

From what you said in the previous comments also in the other issues, my mistake was not thinking about the install stub step which doesn't report errors, and then the binary class is found. Or did I misunderstand that as well? In general, it would be better to avoid such problems in the future, that's why I'd leave a comment in the test. Basically, I didn't understand how that test was supposed to work.

As for the rest, I'm fine with the PR.

File: org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java (hunk 4 above)
Comment (LorenzoBettini) [reply org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.java:129]: Even better ;)

**Review by LorenzoBettini — APPROVED** [review]

**Comment by cdietrich** [comment]
@LorenzoBettini have no longer access to a dev setup. feel free to push a comment with a commit to this brAnch

**Comment by LorenzoBettini** [comment]
> @LorenzoBettini have no longer access to a dev setup. feel free to push a comment with a commit to this brAnch

Yes I can do that, but I still don't understand what was wrong with the previous version of the test.
That's why I asked you. I can put the comment if you explain what the problem was before

**Comment by LorenzoBettini** [comment]
Thank you @cdietrich for adding the comment and fixing this!

**Comment by LorenzoBettini** [comment]
@szarnekow could you please explain where the JavaDerivedStateComputer is used?

**Comment by cdietrich** [comment]
Cause gradle uses the Java source language under the hood to get index entries for Java files in the project

**Comment by LorenzoBettini** [comment]
Ok! That's why it never came into play in my tests and PRs! So Gradle doesn't use Standalone builder. 

So, even more sensible, please try the experiments I said in my comment in the reference projects issue.

**Comment by cdietrich** [comment]
Yes it uses the new builder

org.eclipse.xtext.build.IncrementalBuilder

**Pull request merged**